sql/core/src/main/scala/org/apache/spark/sql/streaming/TwsTester.scala

```diff
@@ -0,0 +1,286 @@
+ *
+ * This class enables unit testing of StatefulProcessor business logic by simulating the
+ * behavior of transformWithState. It processes input rows and returns output rows equivalent
+ * to those that would be produced by the processor in an actual Spark streaming query.
+ *
+ * '''Supported:'''
+ *  - Processing input rows and producing output rows via `test()`.
+ *  - Initial state setup via constructor parameter.
+ *  - Direct state manipulation via `setValueState`, `setListState`, `setMapState`.
+ *  - Direct state inspection via `peekValueState`, `peekListState`, `peekMapState`.
+ *  - Timers in ProcessingTime mode (use `advanceProcessingTime` to fire timers).
+ *  - Timers in EventTime mode (use `eventTimeExtractor` and `watermarkDelayMs` to configure;
+ *    watermark advances automatically based on event times, or use `advanceWatermark` manually).
+ *  - Late event filtering in EventTime mode (events older than the current watermark are dropped).
+ *  - TTL for ValueState, ListState, and MapState (use ProcessingTime mode and
+ *    `advanceProcessingTime` to test expiry).
+ *
+ * '''Testing EventTime Mode:'''
+ *  To test with EventTime, provide `eventTimeExtractor` (a function extracting the event
+ *  timestamp from each input row) and `watermarkDelayMs` (the watermark delay in milliseconds).
+ *  The watermark is computed as `max(event_time_seen) - watermarkDelayMs` and is updated
+ *  automatically after each `test()` call. Late events (with event time older than the current
+ *  watermark) are filtered out before processing, matching production behavior. Timers with
+ *  expiry time <= watermark will fire. You can also manually advance the watermark using
+ *  `advanceWatermark()`.
+ *
+ * '''Use Cases:'''
+ *  - '''Primary''': Unit testing business logic in `handleInputRows` implementations.
+ *  - '''Not recommended''': End-to-end testing or performance testing - use actual Spark
+ *    streaming queries for those scenarios.
+ *
+ * @param processor the StatefulProcessor to test.
+ * @param initialState initial state for each key as a list of (key, state) tuples.
+ * @param timeMode time mode (None, ProcessingTime or EventTime).
+ * @param outputMode output mode (Append, Update, or Complete).
+ * @param realTimeMode whether input rows should be processed one-by-one (separate call to
+ *     handleInputRows) for each input row.
+ * @param eventTimeExtractor function to extract event time from input rows. Required if and
+ *     only if timeMode is EventTime.
+ * @param watermarkDelayMs watermark delay in milliseconds. The watermark is computed as
+ *     `max(event_time) - watermarkDelayMs`. Required if and only if timeMode is EventTime.
+ * @tparam K the type of grouping key.
+ * @tparam I the type of input rows.
+ * @tparam O the type of output rows.
+ * @since 4.0.2
+ */
+class TwsTester[K, I, O](
+    val processor: StatefulProcessor[K, I, O],
+    val initialState: List[(K, Any)] = List(),
```

Comment:
Is it possible we have S in the type list and have `List[(K, S])` as the type of initialState, and provide the default value to `List[(K, Any])()`? I haven't tried so just to see the possibility of enforcing the type for initial state data as well.